<a href="https://colab.research.google.com/github/jeong-wooseok/ML_Study/blob/main/1%EC%A3%BC_1%EA%B0%95_%ED%99%95%EB%A5%A0%EB%B6%84%ED%8F%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1주. 확률분포, 추정과 가설검정, 비모수 검정


## 1강 확률분포


확률변수(Random Variable)는 일정한 확률을 가지고 발생하는 이벤트에 수치가 부여된 변수.
- 확률변수는 X,Y,Z와 같이 대문자로 표기하고, 확률변수의 값은 x,y,z와 같이 소문자표 표기
- 셀 수 있는 확률인 이산형분포, 연속되어 셀수 없는 연속형확률분포로 구분


### 1.1 이산형 분포 확률

확률변수 X가 이산형(Distcrete)인 경우, x 확률이 어떻게 분포되는가
- 확률질량함수로 확률변수 X가 취하는 확률값 x 계산
- 확률값은 0~1사이, 모든 확률의 값은 1
- 베르누이, 이항분포, 음이항분포, 기하분포, 포아송분포, 다항분포, 초기하분포 등이 있다.

[참고사이트](https://lovelydiary.tistory.com/441)



#### 1.1.1 베르누이분포
베르누이 시행은 반복시행 결과가 오직 두가지만 나올수 있는 시행이다.<br>
성공의 횟수를 확률변수 X라 할 때,<br>
확률변수 X는 성공률이 p, 실패확률 (q = 1-p)인 베르누이 분포를 따름<br>
X ~ B(1,p) 로 표기 <br><br>

해당 분포의 확률질량함수, 기대값, 분산을 구하는 공식은 다음과 같다<br>
> $ f(x) = p^x(1-p)^{1-x} \quad\quad x=0,1, \quad\quad   0	\leq q	\leq 1 $<br>
> $ E(x) = p $<br>
> $ Var(X) = pq $<br>


In [1]:
p = 0.3 # 성공할 확률
q = 1-p # 실패할 확률

E = p # 기대값
V = p*q # 분산
print(f'해당분포의 기대값은 {E:.3f}, 분산은 {V:.3f}')

#scipy.stats활용
from scipy.stats import bernoulli
E, V = bernoulli.stats(p)
print(f'scipy 해당분포의 기대값은 {E:.3f}, 분산은 {V:.3f}')


해당분포의 기대값은 0.300, 분산은 0.210
scipy 해당분포의 기대값은 0.300, 분산은 0.210


#### 1.1.2 기하분포

베르누이 시행을 성공할때까지 시행하는 분포. <br>
1) 처음으로 성공할 때까지의 시행횟수를 확률변수 X라 하는 경우 (메인) <br>
2) 처음 성공할때 까지의 실패횟수를 확률변수 X라 하는 경우
확률변수 X는 성공확률 p 를 모수로 갖는 기하분포를 따른다<br>
X ~ G(p) 라 표기<br><br>

- 기하분포의 X는 셀 수는 있지만 값이 무한한 무한확률변수임
- 기하분포는 이산확률분포 중에서 유일하게 무기억성*의 성질을 가짐
 * 무기억성 : 실패를 많이 했다고 해서 실패가 앞으로의 성공할 가능성에 영향을 미치지 않는다는 성질
- E, V = geom.stats(p, moments = 'mv')
- scipy.stats.geom.pmf(x,p), scipy.stats.geom.cdf(x,p)


해당 분포의 확률질량함수, 기대값, 분산을 구하는 공식은 다음과 같다
<br><br>
1) 확률변수 X = 시행횟수
> $ f(x) = pq^{x-1} \quad\quad x=1,2,3... $<br>
 $ E(x) = \frac{1}{p} $<br>
 $ Var(X) = \frac{1}{p^2}$<br>

<br>


2) 확률변수 X = 실패횟수
> $ f(x) = pq^{x} \quad\quad x=0,1,2... $<br>
 $ E(x) = \frac{q}{p} $ <br>
 $ Var(X) = \frac{q}{p^2}$ <br>




Q) 하나의 주사위를 세번 던질 때 세번째 시행에서 앞면 숫자가 6이 나올 확률은?


In [9]:
from scipy.stats import geom
n = 3
p = 1/6
q= 1-p

case = 1
if case == 1:
  x = n
  E = 1 / p
  V = q / p**2
  E2, V2 = geom.stats(p, moments = 'mv') # mean, variance 
  pkg_pmf, pkg_cdf = geom.pmf(x,p), geom.cdf(x,p)
  cdf = 0
  for i in range(1,x+1):
    cdf += q**(i-1) * p
  
elif case == 2:
  x = n - 1
  E = q / p
  V = q / p**2
  pmf = q**x * p
  cdf = 0
  for i in range(1,x+1):
    cdf += q**(i-1) * p

# 기대값과 분산 계산하기
print(f'[수기] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 
print(f'[라이브러리리] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 

# 확률질량함수 계산하기
print(f'[수기] 확률질량함수 (pmf) : {pmf:.3f}')
print(f'[라이브러리] 확률질량함수 (pmf) : {pkg_pmf:.3f} \n')

# 누적확률질량함수 계산하기 : 2번째 이하로 이길 확률은? 
print(f'[수기] 누적확률질량함수 (cdf) : {cdf:.3f}')
print(f'[라이브러리] 누적확률질량함수 (cdf) : {pkg_cdf:.3f} \n')

print(f'하나의 주사위를 세번 던질 때 세번째 시행에서 앞면 숫자가 6이 나올 확률은 {pmf*100:.1f}% 이다')


[수기] 해당분포 기대값은 6.000, 분산은 30.000
[라이브러리리] 해당분포 기대값은 6.000, 분산은 30.000
[수기] 확률질량함수 (pmf) : 81.002
[라이브러리] 확률질량함수 (pmf) : 0.116 

[수기] 누적확률질량함수 (cdf) : 0.421
[라이브러리] 누적확률질량함수 (cdf) : 0.421 

하나의 주사위를 세번 던질 때 세번째 시행에서 앞면 숫자가 6이 나올 확률은 8100.2% 이다


####  1.1.3 이항분포

베르누이 시행을 "독립적으로 n번 반복시행"했을 때 분포

성공의 횟수를 확률변수 X라하면<br>
확률변수 X는 시행횟수 n번과 성공확률 p, 실패확률 (q=1-p)를 모수로 하는 <br>이항분포를 따른다.<br>
X ~ B(n,p)로 표기<br>

- 이항분포 B(n,p)는 n>20, np>=5, n(1-p)>=5일 때 정규분포 N(np, npq)로 근사한다.
- p <= 0.1, np= 0.1 ~ 10, n >= 50 일때 포아송분포 P(np)로 근사한다.
- scipy.stats.binom.pmf(x,n,p), scipy.stats.binom.cdf(x,n,p)


해당 분포의 확률질량함수, 기대값, 분산을 구하는 공식은 다음과 같다
> $ f(x) = _n C_x p^x(1-p)^{n-x} \quad\quad x=0,1,...,n $<br>
> $ E(x) = np $<br>
> $ Var(X) = npq $<br>


Q) 근로자가 내년에 회사에서 일하지 않을 확률이 0.1이라고 추정한 경우,<br>
시간제 근로자 3명을 무작위로 뽑았을 때, 그 중 한명이 금년에 회사를 떠날 확률은?


In [3]:
from scipy.stats import binom
import numpy as np

# combination 함수정의 : 수기로 계산시 확률계산용
fac = np.math.factorial
def combi(n,x):
  result = fac(n)/fac(x)/fac(n-x)
  return result

In [4]:
n = 3
x = 1
p = 0.1

E = n*p
V = n*p*(1-p)
print(f'해당분포의 기대값은 {E:.3f}, 분산은 {V:.3f}')

# 확률질량함수 계산
pmf = combi(n,x) * p**x * (1-p)**(n-x)
print(f'해당분포의 확률질량함수(pmf):{pmf:.3f}')

# 누적확률질량 함수 계산
cdf = 0
for i in range(0,x+1):
  cdf += combi(n,x) * p**x * (1-p)**(n-x)
print(f'해당분포의 누적확률질량함수(cdf):{binom.cdf(x,n,p):.3f}')
 

해당분포의 기대값은 0.300, 분산은 0.270
해당분포의 확률질량함수(pmf):0.243
해당분포의 누적확률질량함수(cdf):0.972


무작위로 3명 뽑을 때 그중 한명이 금년에 회사를 떠날 확률은 0.243임

#### 1.1.4 음이항분포

베르누이 시행을 "독립적으로 k번 성공할 때까지 반복시행"


베르누이 시행을 "독립적으로 k번 성공할 때까지 반복시행"했을 때 <br>
1) 시행횟수를 확률변수 X라 하거나 (일반적으로 분석가가 관심을 가짐)<br>
2) 실패횟수를 확률변수 X라 하는 경우<br>
확률변수 X는 성공횟수 k와 성공확률 p 를 모수로 갖는 음이항 분포를 따른다<br>
X ~ NB(k,p) 라 표기<br>

- 음이항분포의 X는 셀 수는 있지만 값이 무한한 무한확률변수.
- 음이항분포에서 k=1이면, 음이항분포는 기하분포가 됨 
- E, V = scipy.stats.nbinom.stats(k,p)
- scipy.stats.nbinom.pmf(x,k,p), scipy.stats.nbinom.cdf(x,k,p)


해당 분포의 확률질량함수, 기대값, 분산을 구하는 공식은 다음과 같다<br>
<br><br>
1) 확률변수 X = 시행횟수
> $ f(x) = _{x-1} C_{k-1} p^k(1-p)^{x-k} \quad\quad x=k,k+1,...,\quad\quad k>0 $<br>
 $ E(x) = \frac{kq}{p} $<br>
 $ Var(X) = \frac{kq}{p^2}$<br>

<br>


2) 확률변수 X = 실패횟수
> $ f(x) = _{x+k-1} C_{k-1} p^k(1-p)^x \quad\quad x=0,1,2,...,\quad\quad k>0 $ <br>
 $ E(x) = \frac{kq}{p} $ <br>
 $ Var(X) = \frac{kq}{p^2}$ <br>


Q) A가 승리할 확률이 0.3 이라면 5번째 경기에서 2번째로 이길 확률은? 


In [5]:
from scipy.stats import nbinom
n = 5 # 총 시행횟수
k = 2 # 총 성공횟수
p = 0.3 # 성공확률
q = 1-p # 실패확률

case = 2
if case ==1: # 확률변수 X가 k번 성공할 때까지 시행횟수 x인 경우
  x = n # **확률변수
  E = k / p
  V = k*q / p**2
  pmf = combi(x-1,k-1) * p**k * q**(x-k)
  cdf=0
  for i in range(k,x+1): # 확률변수의 확률값들을 합함
    cdf += combi(i-1,k-1) * p**k * q**(1-k)
elif case ==2: # 확률변수 X가 k번 성공할 때까지의 실패횟수 x인 경우
  x = n-k # **확률변수
  E = k*q / p
  V = k*q / p**2
  pmf = combi(x+k-1, k-1) * p**k * q**(x)
  E2, V2 = nbinom.stats(k,p)
  pkg_pmf, pkg_cdf = nbinom.pmf(x,k,p), nbinom.cdf(x,k,p)
  cdf = 0
  for i in range(0, x+1): # 확률변수의 확률값들을 합함
    cdf += nbinom.pmf(i,k,p)

# 기대값과 분산 계산하기 
print(f'[수기] 해당분포의 기대값은{E:.3f}, 분산은 {V:.3f}')
print(f'[라이브러리] 해당분포의 기대값은{E2:.3f}, 분산은 {V2:.3f} \n')

# 확률질량함수 계산하기
print(f'[수기] 확률질량함수 (pmf) : {pmf:.3f}')
print(f'[라이브러리] 확률질량함수 (pmf) : {pkg_pmf:.3f} \n')

# 누적확률질량함수 계산하기 : 2번째 이하로 이길 확률은? 
print(f'[수기] 누적확률질량함수 (cdf) : {cdf:.3f}')
print(f'[라이브러리] 누적확률질량함수 (cdf) : {cdf:.3f} \n')

print(f'5번째 경기에서 2번째로 이길 확률은 {pmf*100:.1f}% 이다')



[수기] 해당분포의 기대값은4.667, 분산은 15.556
[라이브러리] 해당분포의 기대값은4.667, 분산은 15.556 

[수기] 확률질량함수 (pmf) : 0.123
[라이브러리] 확률질량함수 (pmf) : 0.123 

[수기] 누적확률질량함수 (cdf) : 0.472
[라이브러리] 누적확률질량함수 (cdf) : 0.472 

5번째 경기에서 2번째로 이길 확률은 12.3% 이다


#### 1.1.5 초기하분포
비복원추출에 적용되는 분포로 첫 시도의 결과에 따라 두번째 시도의 상황이 달라지므로 시도들이 종속관계에 있다 <br>


- 크기 N 의 유한모집단 중 크기 n의 확률표본을 뽑을 경우, <br>
  N개 중 k개는 성공으로, 나머지 (N-k)개는 실패로 분류하여 비복원으로 뽑을 때  <br>
  성공 횟수를 X라 하면,   확률변수 X는 N, k, n을 모수로 갖는 초기하분포를 따른다고 한다.<br>
- X ~ HG(N, k, n)<br>
- scipy.stats.hypergeom.pmf(x, N, k, n), scipy.stats.hypergeom.cdf(i, N, k, n)

해당 분포의 확률질량함수, 기대값, 분산을 구하는 공식은 다음과 같다
<br><br>
> $ f(x) = \frac{ _{k}C_x *  _{(N-k)}C_{(n-k)}}{_NC_n} \quad\quad x=0,1...,n $<br>
 $ E(x) = \frac{nk}{N} $<br>
 $ Var(X) = \frac{nk}{N}\frac{n-k}{N}\frac{N-n}{N-1}$<br>




Q) 상자속에 빨간 공이 90개, 파란 공이 10개 들어있다. 
임의로 1개씩 두번 꺼내고 다시 넣지 않을 때 1개가 파란 공이 될 확률은?


In [6]:
from scipy.stats import hypergeom

N = 100 # 모집단 수 (총 공개수)
k = 10 # 성공요소 수 (총 파란공 개수)
n = 2 # 시행횟수 (공 선택개수)
x = 1 # 확인하고자 하는 사건 (1개가 파란공) ** 확률변수
P = k/N # 모비율

# 기대값과 분산 구하기
E = n * p
V = n * p * (1-p) * ((N-n) / (N-1))
E2, V2 = hypergeom.stats(N,k,n)

# 기대값과 분산 계산하기
print(f'[수기] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 
print(f'[라이브러리리] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 

# 확률질량함수/확률누계함수 구하기
pmf = combi(k,x) * combi(N-k,n-x) / combi(N,n)
cdf = 0

pkg_pmf, pkg_cdf = hypergeom.pmf(x,N,k,n), hypergeom.cdf(x,N,k,n)

# 확률질량함수 계산하기 : 임의로 1개씩 두번 꺼내고 다시 넣지 않을 때 1개가 파란 공이 될 확률은?
print(f'[수기] 확률질량함수 (pmf) : {pmf:.3f}')
print(f'[라이브러리] 확률질량함수 (pmf) : {pkg_pmf:.3f} \n')

# 누적확률질량함수 계산하기 : 파란공이 1개 이하일 확률은
# print(f'[수기] 누적확률질량함수 (cdf) : {cdf:.3f}')
print(f'[라이브러리] 누적확률질량함수 (cdf) : {pkg_cdf:.3f} \n')



[수기] 해당분포 기대값은 0.600, 분산은 0.416
[라이브러리리] 해당분포 기대값은 0.600, 분산은 0.416
[수기] 확률질량함수 (pmf) : 0.182
[라이브러리] 확률질량함수 (pmf) : 0.182 

[라이브러리] 누적확률질량함수 (cdf) : 0.991 



In [7]:
# 초기하분포의 이항분포 근사
from scipy.stats import norm
b_result = binom.pmf(x,n,k/N)
h_result = hypergeom.pmf(x,N,k,n)
print(f'초기하분포로 계산하면 {h_result:.3f}, 이항분포로 근사계산하면 {b_result:.3f}')

초기하분포로 계산하면 0.182, 이항분포로 근사계산하면 0.180


#### 1.1.6 포아송분포


단위시간, 단위면적 또는 단위 공간 내에서 발생하는 어떤 사건의 횟수를 확률변수 X라할때,<br>
- 확률변수 X 는 Lambda(단위시간, 단위면적/공간 내에 발생하는 사건의 평균값)를 <br>모수로 갖는 포아송 분포를 따른다.<br>
- 포아송분포의 X는 셀수는 있지만 값이 무한한 무한 확률변수이다


Q) 주말 저녁 시간 당 평균 6명이 응급실에 올 경우, 어떤 주말 저녁 30분 내 4명이 도착할 확률은?




In [8]:
from scipy.stats import poisson

x = 4 # 구간 내 사건 횟수 ** 확률변수 
lam = 3 # 구간 내 평균 (=lambda)

# 기대값과 분산 계산하기
E= lam
V = lam
E2, V2 = poisson.stats(lam, moments='mv')

# 기대값과 분산 계산하기
print(f'[수기] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 
print(f'[라이브러리리] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}\n') 

# 확률질량함수 계산하기

pmf = lam**x + np.exp(-lam) / np.math.factorial(x)

cdf = 0
for i in range(0, x+1):
  cdf += lam**x + np.exp(-lam) / np.math.factorial(x)

pkg_pmf, pkg_cdf = poisson.pmf(x,lam), poisson.cdf(x,lam)

# 확률질량함수 계산하기 :주말 저녁 시간 당 평균 6명이 응급실에 올 경우, 어떤 주말 저녁 30분 내 4명이 도착할 확률은?
print(f'[수기] 확률질량함수 (pmf) : {pmf:.3f}')
print(f'[라이브러리] 확률질량함수 (pmf) : {pkg_pmf:.3f} \n')

# 누적확률질량함수 계산하기 : 주말 저녁 시간 당 평균 6명이 응급실에 올 경우, 어떤 주말 저녁 30분 내 4명 이하로 도착할 확률은?
print(f'[수기] 누적확률질량함수 (cdf) : {cdf:.3f}')
print(f'[라이브러리] 누적확률질량함수 (cdf) : {pkg_cdf:.3f} \n')


[수기] 해당분포 기대값은 3.000, 분산은 3.000
[라이브러리리] 해당분포 기대값은 3.000, 분산은 3.000

[수기] 확률질량함수 (pmf) : 81.002
[라이브러리] 확률질량함수 (pmf) : 0.168 

[수기] 누적확률질량함수 (cdf) : 405.010
[라이브러리] 누적확률질량함수 (cdf) : 0.815 



In [10]:
from scipy.stats import norm

x = 10
lam = 20
E = V = lam

sigma = np.sqrt(V) #포아송분포의 표준편차
zstat1 = (x-0.5-lam)/sigma
zstat2 = (x+0.5-lam)/sigma
n_result = norm.cdf(zstat2) -norm.cdf(zstat1)
p_result = poisson.pmf(x,lam)

print('포아송분포로 계산하면 {:3f} 정규분포로 계산하면 {:.3f}'.format(p_result,n_result))

포아송분포로 계산하면 0.005816 정규분포로 계산하면 0.007


### 1.2 연속형 확률분포
확률변수가 가질 수 있는 값이 연속적인 실수여서 셀 수 없는 경우의 분포
- 확률변수 X의 값은 확률밀도함수를 이용하여 계산
- 확률밀도함수를 적분함으로써 그래프 아래의 면적인 확률값 계산 (면적의 총합 =1)
- 균일분포, 정규분포, 지수분포, 카이제곱분포, t분포, F분포 등이 있음

#### 1.2.1 균일분포

구간 (a,b)에서 확률변수 X가 균일하게 분포<br>
X ~ U(a, b)<br>
- scipy.stats.norm.pdf(x,lam)

> $f(x) = \frac{1}{b-a},\quad\quad a<x<b $<br>
> $E(X) = \mu $<br>
> $Var(X) = σ^2 $<br>

확률변수 X가 (5,15)에서 균일분포를 따를 때 12와 115 사이의 확률은?

In [12]:
[range(12,15)]

[range(12, 15)]

In [13]:
# 균일분포
a = 5 # 확률밀도함수 범위
b = 15 # 확률밀도함수 범위
range = [(12,15)]
p = 1/(b-a) #확률밀도함수

E = (b+a)/2
V = (b-a)**2 / 12
cdf = 0
for (x1,x2) in range:
  cdf += (x2-x1) * p 
print(f'[수기] 해당분포 기대값은 {E:.3f}, 분산은 {V:.3f}') 
print(f'[수기] 범위내 누적확률은 {cdf:.3f}') 

[수기] 해당분포 기대값은 10.000, 분산은 8.333
[수기] 범위내 누적확률은 0.300


```
Q) 버스가 오전 7시부터 15분 간격으로 정류장을 출발한다.
한 승객이 이 정거장에 도착하는 시간은 


#### 1.2.2 정규분포, 표준 정규분포

#### 1.2.3 지수분포

#### 1.2.4 감마분포

#### 1.2.5 카이제곱분포

#### 1.2.6 t분포

#### 1.2.7 F분포